test/services/task_list_service_test.dart

```diff
@@ -1,47 +1,79 @@
-  late IdentityService identityService;
-  late SyncService syncService;
-  late TaskListService taskListService;
+  late DeviceTaskList device;
 
   setUp(() async {
-    database = await databaseFactoryMemory.openDatabase('');
-    keyValueRepository = KeyValueRepository(database);
-    identityService = IdentityService(keyValueRepository);
-    syncService = SyncService(keyValueRepository);
-    taskListService =
-        TaskListService(keyValueRepository, identityService, syncService);
+    device = await DeviceTaskList.create(name: 'device');
+  });
+
+  test('create task list', () async {
+    final id = Uuid().v4();
+    await device.taskListService
+        .upsertTaskList(TaskList(id: id, title: 'list1'));
+    final taskListRecordMap = await device.taskListService.taskLists;
+    expect(taskListRecordMap.map((e) => e.id!), [id]);
+    expect(taskListRecordMap.first.title, 'list1');
   });
 
   test('should store and retrieve tasks after first task is deleted', () async {
-    final taskTitle = 'Drink a cold cat';
-
-    await taskListService.upsert(
-      Task(
-        title: 'Catch a cat falling from the sky',
-        taskListID: '1',
-      ),
-    );
-    final task = (await taskListService.tasks).first;
-    await taskListService.remove(task);
-    await taskListService.upsert(Task(title: taskTitle, taskListID: '1'));
-    final tasks = await taskListService.tasks;
-
-    expect(tasks.length, equals(1));
-    expect(tasks.first.title, equals(taskTitle));
+    const taskTitle1 = 'Catch a cat falling from the sky';
+    const taskTitle2 = 'Drink a cold cat';
+
+    await device.taskListService
+        .upsertTaskList(TaskList(id: 'id1', title: 'list1'));
+    await device.taskListService.upsertTask('id1', Task(title: taskTitle1));
+    var tasks = (await device.taskListService.allTasks).toList();
+    expect(tasks.length, 1);
+    expect(tasks.first.title, taskTitle1);
+    final taskId = tasks.first.id!;
+    expect(taskId, isNot(null));
+
+    await device.taskListService.removeTask('id1', taskId);
+    tasks = (await device.taskListService.allTasks).toList();
+    expect(tasks.isEmpty, true);
+
+    await device.taskListService.upsertTask('id1', Task(title: taskTitle2));
+    tasks = (await device.taskListService.allTasks).toList();
+
+    expect(tasks.length, 1);
+    expect(tasks.first.title, taskTitle2);
+  });
+
+  test('should retrieve tasks with allTaskRecords', () async {
+    final task1 = Task(title: 'Catch a cat falling from the sky');
+    final task2 = Task(title: 'Drink a cold cat');
+    await device.taskListService
+        .upsertTaskList(TaskList(id: 'id1', title: 'list1'));
+    await device.taskListService.upsertTask('id1', task1);
+    await device.taskListService.upsertTask('id1', task2);
+
+    final allTaskRecords =
+        (await device.taskListService.allTaskRecords).toList();
+    expect(allTaskRecords.length, 2);
+    expect(allTaskRecords.map((v) => v.task).toSet(), {task1, task2});
```

Comment:
It's a set, it doesn't have an order.
I do this to check that the tasks are contained and that they are the only tasks.

Reply:
Ok, I wasn't sure how the comparison behaves.
Let's use the [containsAll](https://api.dart.dev/stable/2.13.4/dart-core/Set/containsAll.html) method here to make this more explicit:

`allTaskRecords.toSet().containsAll({task1, task2})`

(With allTaskRecords being a [list](https://api.dart.dev/stable/2.13.4/dart-core/List-class.html), `.toSet()` should work without the map.)

Reply:
The `map` is necessary because `allTaskRecords` is of type `List<TaskRecord>`.
I think using sets for this kind of comparison is very readable and intuitive, but I agree that the previous line 58 is unnecessary and I will remove it. Using `containsAll` only checks if all elements are contained, but not if they are the only elements (in combination with line 58 it would be equivalent though).

Reply:
See same problem as above